# House price prediction using multiple regression analysis

# Part 2: Regression Models

The following notebook presents a thought process of predicting a continuous variable through Machine Learning methods. More specifically, we want to predict house prices based on multiple features using regression analysis. 

As an example, we will use a dataset of house sales in King County, where Seattle is located.

In this second notebook we will apply multiple regression models. We will talk about model complexity and how one can select the best predictive model using a validation set or cross-validation techniques.

## 1. Preparation

As in Part 1, Let's first load the libraries and the dataset

In [1]:
import numpy as np # NumPy is the fundamental package for scientific computing

import pandas as pd # Pandas is an easy-to-use data structures and data analysis tools
pd.set_option('display.max_columns', None) # To display all columns

import matplotlib.pyplot as plt # Matplotlib is a python 2D plotting library
%matplotlib inline 
# A magic command that tells matplotlib to render figures as static images in the Notebook.

import seaborn as sns # Seaborn is a visualization library based on matplotlib (attractive statistical graphics).
sns.set_style('whitegrid') # One of the five seaborn themes
import warnings
warnings.filterwarnings('ignore') # To ignore some of seaborn warning msg

from scipy import stats

from sklearn import linear_model # Scikit learn library that implements generalized linear models
from sklearn import neighbors # provides functionality for unsupervised and supervised neighbors-based learning methods
from sklearn.metrics import mean_squared_error # Mean squared error regression loss
from sklearn import preprocessing # provides functions and classes to change raw feature vectors

from math import log

In [2]:
data = pd.read_csv("kc_house_data.csv", parse_dates = ['date']) # load the data into a pandas dataframe
data.head(2) # Show the first 2 lines

id       date     price  bedrooms  bathrooms  sqft_living  \
0  7129300520 2014-10-13  221900.0         3       1.00         1180   
1  6414100192 2014-12-09  538000.0         3       2.25         2570   

   sqft_lot  floors  waterfront  view  condition  grade  sqft_above  \
0      5650     1.0           0     0          3      7        1180   
1      7242     2.0           0     0          3      7        2170   

   sqft_basement  yr_built  yr_renovated  zipcode      lat     long  \
0              0      1955             0    98178  47.5112 -122.257   
1            400      1951          1991    98125  47.7210 -122.319   

   sqft_living15  sqft_lot15  
0           1340        5650  
1           1690        7639

### Data Cleaning

Let's reduce the dataset by dropping columns that won't be used during the analysis.

In [3]:
data.drop(['id', 'date', 'zipcode'], axis = 1, inplace = True)

### Data Transformation

Following the correlation analysis in Part 1, let's create some new variables in our dataset. 

In [4]:
data['basement_present'] = data['sqft_basement'].apply(lambda x: 1 if x > 0 else 0) # Indicate whether there is a basement or not
data['renovated'] = data['yr_renovated'].apply(lambda x: 1 if x > 0 else 0) # 1 if the house has been renovated

### Encode categorical variables using dummies

A Dummy variable is an artificial variable created to represent an attribute with two or more distinct categories/levels. In this example, we will analyse *bedrooms* and *bathrooms* as continuous and therefore will only encode the following:

* *floors*
* *view*
* *condition* and
* *grade*

In [5]:
categorial_cols = ['floors', 'view', 'condition', 'grade']

for cc in categorial_cols:
    dummies = pd.get_dummies(data[cc], drop_first=False)
    dummies = dummies.add_prefix("{}#".format(cc))
    data.drop(cc, axis=1, inplace=True)
    data = data.join(dummies)
    
data.dtypes

price               float64
bedrooms              int64
bathrooms           float64
sqft_living           int64
sqft_lot              int64
waterfront            int64
sqft_above            int64
sqft_basement         int64
yr_built              int64
yr_renovated          int64
lat                 float64
long                float64
sqft_living15         int64
sqft_lot15            int64
basement_present      int64
renovated             int64
floors#1.0          float64
floors#1.5          float64
floors#2.0          float64
floors#2.5          float64
floors#3.0          float64
floors#3.5          float64
view#0              float64
view#1              float64
view#2              float64
view#3              float64
view#4              float64
condition#1         float64
condition#2         float64
condition#3         float64
condition#4         float64
condition#5         float64
grade#1             float64
grade#3             float64
grade#4             float64
grade#5             

### Split the data

We will split the dataframe into training and testing data using a 80%/20% ratio

In [6]:
from sklearn.cross_validation import train_test_split
train_data, test_data = train_test_split(data, train_size = 0.8, random_state = 10)

## 2. Regression Models

In this section, we will train numerous regression models on the train data (e.g., simple linear regression, lasso, nearest neighbor) and evaluate their performance using Root Mean Squared Error (RMSE) on the test data.

### 2.1 Simple Linear Regression

Let's first predict house prices using simple (one input) linear regression.

In [7]:
# A function that take one input of the dataset and return the RMSE (of the test data), and the intercept and coefficient
def simple_linear_model(train, test, input_feature):
    regr = linear_model.LinearRegression() # Create a linear regression object
    regr.fit(train.as_matrix(columns = [input_feature]), train.as_matrix(columns = ['price'])) # Train the model
    RMSE = mean_squared_error(test.as_matrix(columns = ['price']), 
                              regr.predict(test.as_matrix(columns = [input_feature])))**0.5 # Calculate the RMSE on test data
    return RMSE, regr.intercept_[0], regr.coef_[0][0]

let's create a simple linear regression model using *sqft_living* as input and calculate the RMSE on the test data.

In [8]:
RMSE, w0, w1 = simple_linear_model(train_data, test_data, 'sqft_living')
print 'RMSE for sqft_living is: %s ' %RMSE
print 'intercept is: %s' %w0
print 'coefficient is: %s' %w1

RMSE for sqft_living is: 268279.643883 
intercept is: -36738.1773464
coefficient is: 277.36412987


Similarly, we can run the same test on all the features in the dataset and assess which one would be the best estimator of house price using just a single linear regression model.

In [9]:
input_list = data.columns.values.tolist() # list of column name
input_list.remove('price')
simple_linear_result = pd.DataFrame(columns = ['feature', 'RMSE', 'intercept', 'coefficient'])

# loop that calculate the RMSE of the test data for each input 
for p in input_list:
    RMSE, w1, w0 = simple_linear_model(train_data, test_data, p)
    simple_linear_result = simple_linear_result.append({'feature':p, 'RMSE':RMSE, 'intercept':w0, 'coefficient': w1}
                                                       ,ignore_index=True)
simple_linear_result.sort_values('RMSE').head(10) # display the 10 best estimators

feature           RMSE      intercept   coefficient
2     sqft_living  268279.643883     277.364130 -3.673818e+04
5      sqft_above  304131.310592     266.306764  6.461714e+04
11  sqft_living15  320686.541323     314.359911 -8.502590e+04
1       bathrooms  324082.781919  246523.891877  1.863279e+04
21         view#0  356019.001320 -435033.777431  9.322014e+05
6   sqft_basement  357843.745395     258.126523  4.642966e+05
40       grade#11  357964.423743  965286.415396  5.222663e+05
39       grade#10  360773.700418  556992.601325  5.107024e+05
0        bedrooms  361295.375626  117579.891853  1.436815e+05
9             lat  365041.433662  814499.981062 -3.819789e+07

When using simple linear regression, *sqft_living* provides the smallest test error estimate of house price for the dataset considered. 

### 2.2 Multiple Regression

Now let's try to predict *price* using multiple features. We can modify the simple linear regression function above to take multiple features as input.

In [10]:
# A function that take multiple features as input and return the RMSE (of the test data), and the  intercept and coefficients
def multiple_regression_model(train, test, input_features):
    regr = linear_model.LinearRegression() # Create a linear regression object
    regr.fit(train.as_matrix(columns = input_features), train.as_matrix(columns = ['price'])) # Train the model
    RMSE = mean_squared_error(test.as_matrix(columns = ['price']), 
                              regr.predict(test.as_matrix(columns = input_features)))**0.5 # Calculate the RMSE on test data
    return RMSE, regr.intercept_[0], regr.coef_ 

Let's try with a few examples:

In [11]:
print 'RMSE: %s, intercept: %s, coefficients: %s' %multiple_regression_model(train_data, 
                                                                             test_data, ['sqft_living','bathrooms','bedrooms'])
print 'RMSE: %s, intercept: %s, coefficients: %s' %multiple_regression_model(train_data, 
                                                                             test_data, ['sqft_above','view#0','bathrooms'])
print 'RMSE: %s, intercept: %s, coefficients: %s' %multiple_regression_model(train_data, 
                                                                             test_data, ['bathrooms','bedrooms'])
print 'RMSE: %s, intercept: %s, coefficients: %s' %multiple_regression_model(train_data, 
                                                                             test_data, ['view#0','grade#12','bedrooms','sqft_basement'])
print 'RMSE: %s, intercept: %s, coefficients: %s' %multiple_regression_model(train_data, 
                                                                             test_data, ['sqft_living','bathrooms','view#0'])

RMSE: 264872.283555, intercept: 81100.9596775, coefficients: [[   306.15090562   7913.53847651 -57658.90103459]]
RMSE: 282802.364962, intercept: 303531.292188, coefficients: [[  1.98928518e+02  -3.17760670e+05   7.92684094e+04]]
RMSE: 323412.269276, intercept: -18432.9057072, coefficients: [[ 235300.89998266   18030.65120532]]
RMSE: 320893.658432, intercept: 507958.889992, coefficients: [[ -3.31957675e+05   1.35299763e+06   8.56179339e+04   1.24073575e+02]]
RMSE: 260210.048524, intercept: 205076.198921, coefficients: [[    258.00653033    -244.89749378 -223120.61245789]]


We can also try to fit a higher-order polynomial on the input. For example, we can try to fit a qudratic function on *sqft_living*

In [12]:
train_data['sqft_living_squared'] = train_data['sqft_living'].apply(lambda x: x**2) # create a new column in train_data
test_data['sqft_living_squared'] = test_data['sqft_living'].apply(lambda x: x**2) # create a new column in test_data
print 'RMSE: %s, intercept: %s, coefficients: %s' %multiple_regression_model(train_data, 
                                                                             test_data, ['sqft_living','sqft_living_squared'])

RMSE: 246063.959207, intercept: 172106.560106, coefficients: [[  9.31721994e+01   3.37454932e-02]]


While we can get better performance than simple linear models, a few problems remain. 

* First, we don't know which feature to select. Obviously some combinations of features will yield smaller RMSE on the test set.
* Second, we don't know how many features to select. This is because the more features we incorporate in the train model, the more overfit we get on the train data, resulting in higher error on the test data.

One solution would be to test multiple features combinations (all?) and keep the solution with the smallest error value calculated on the test data. However, this is an overly optimistic approach, since the model complexity is selected to minimize the test error (error is biased). A more sophisticated approach is to use two sets for testing our models, a.k.a: a validation set and a test set. We select model complexity to minimize error on the validation set and approximate the generalization error based on the test set.

Going through all subsets of features combinations is most often computationally infeasible. For example, having 30 features yield more than 1 billion combinations. Another approach is to use a greedy technique like a forward stepwise algorithm where the best estimator feature is added to the set of already selected features at each iteration. For example, let's pretend that the best single estimator is *sqft_living*. In the 2nd step of the greedy algorithm, we test all the remaining features one by one in combinations with *sqft_living* (e.g., *sqft_living* and *bedrooms*, *sqft_living* and *waterfront*, etc) and select the best combination using training error. At the end, we select the model complexity (number of features) using the validation error and estimate the generalization error using the test set.

Let's try this method.

In [13]:
# we're first going to add more features into the dataset.

# sqft_living cubed
train_data['sqft_living_cubed'] = train_data['sqft_living'].apply(lambda x: x**3) 
test_data['sqft_living_cubed'] = test_data['sqft_living'].apply(lambda x: x**3) 

# bedrooms_squared: this feature will mostly affect houses with many bedrooms.
train_data['bedrooms_squared'] = train_data['bedrooms'].apply(lambda x: x**2) 
test_data['bedrooms_squared'] = test_data['bedrooms'].apply(lambda x: x**2)

# bedrooms times bathrooms gives what's called an "interaction" feature. It is large when both of them are large.
train_data['bed_bath_rooms'] = train_data['bedrooms']*train_data['bathrooms']
test_data['bed_bath_rooms'] = test_data['bedrooms']*test_data['bathrooms']

# Taking the log of squarefeet has the effect of bringing large values closer together and spreading out small values.
train_data['log_sqft_living'] = train_data['sqft_living'].apply(lambda x: log(x))
test_data['log_sqft_living'] = test_data['sqft_living'].apply(lambda x: log(x))

train_data.shape

(17290, 49)

In [14]:
# split the train_data to include a validation set (train_data2 = 60%, validation_data = 20%, test_data = 20%)
train_data_2, validation_data = train_test_split(train_data, train_size = 0.75, random_state = 50)

In [15]:
# A function that take multiple features as input and return the RMSE (of the train and validation data)
def RMSE(train, validation, features, new_input):
    features_list = list(features)
    features_list.append(new_input)
    regr = linear_model.LinearRegression() # Create a linear regression object
    regr.fit(train.as_matrix(columns = features_list), train.as_matrix(columns = ['price'])) # Train the model
    RMSE_train = mean_squared_error(train.as_matrix(columns = ['price']), 
                              regr.predict(train.as_matrix(columns = features_list)))**0.5 # Calculate the RMSE on train data
    RMSE_validation = mean_squared_error(validation.as_matrix(columns = ['price']), 
                              regr.predict(validation.as_matrix(columns = features_list)))**0.5 # Calculate the RMSE on train data
    return RMSE_train, RMSE_validation 

In [16]:
input_list = train_data_2.columns.values.tolist() # list of column name
input_list.remove('price')

# list of features included in the regression model and the calculated train and validation errors (RMSE)
regression_greedy_algorithm = pd.DataFrame(columns = ['feature', 'train_error', 'validation_error'])  
i = 0
temp_list = []

# a while loop going through all the features in the dataframe
while i < len(train_data_2.columns)-1:
    
    # a temporary dataframe to select the best feature at each iteration
    temp = pd.DataFrame(columns = ['feature', 'train_error', 'validation_error'])
    
    # a for loop to test all the remaining features
    for p in input_list:
        RMSE_train, RMSE_validation = RMSE(train_data_2, validation_data, temp_list, p)
        temp = temp.append({'feature':p, 'train_error':RMSE_train, 'validation_error':RMSE_validation}, ignore_index=True)
        
    temp = temp.sort_values('train_error') # select the best feature using train error
    best = temp.iloc[0,0]
    temp_list.append(best)
    regression_greedy_algorithm = regression_greedy_algorithm.append({'feature': best, 
                                                  'train_error': temp.iloc[0,1], 'validation_error': temp.iloc[0,2]}, 
                                                 ignore_index=True) # add the feature to the dataframe
    input_list.remove(best) # remove the best feature from the list of available features
    i += 1
regression_greedy_algorithm

feature    train_error  validation_error
0   sqft_living_squared  249721.440480     268747.213950
1                   lat  229419.062478     252069.707424
2            waterfront  219240.069871     241979.268406
3                view#0  214598.840402     235006.341864
4         sqft_living15  212061.440586     225553.945689
5              yr_built  209106.952300     224631.813173
6              grade#13  207150.588541     224190.105488
7             bathrooms  205130.520723     220147.077763
8              grade#11  203469.225546     217463.284605
9              grade#10  201188.054657     212822.430513
10             grade#12  198206.357564     206889.961277
11              grade#9  194396.120835     200131.271456
12              grade#8  191829.045018     196721.407209
13              grade#7  190795.419372     195123.929145
14               view#4  189824.261708     195545.525439
15          condition#5  189102.542859     194989.957808
16    sqft_living_cubed  188395.497956     215653.150504
17                 long  187947.022313     214924.040846
18          condition#4  187536.098211     214809.627580
19         yr_renovated  187080.601484     215263.306150
20            renovated  186706.152536     214997.564561
21           floors#1.0  186511.550441     215307.790744
22               view#2  186357.363644     215418.323430
23       bed_bath_rooms  186217.158957     214035.613580
24           floors#2.5  186102.702791     213924.003501
25           sqft_lot15  185989.051869     213464.047937
26     basement_present  185883.921066     214876.863444
27           sqft_above  185805.566668     217453.155954
28           floors#3.0  185776.782696     217405.452628
29              grade#5  185750.456583     217371.361223
30           floors#3.5  185730.554156     217386.548045
31        sqft_basement  185716.964969     221367.775692
32     bedrooms_squared  185704.784672     221217.748219
33             sqft_lot  185693.726064     221197.624977
34              grade#3  185682.676406     221305.890576
35             bedrooms  185679.964784     221698.809200
36      log_sqft_living  185677.991805     223483.928825
37          condition#2  185676.072928     223525.134348
38              grade#6  185674.912036     223827.924724
39               view#1  185674.033300     223822.398963
40           floors#1.5  185673.833295     223868.008481
41              grade#4  185673.830091     223870.720189
42           floors#2.0  185673.824393     223870.018899
43               view#3  185673.823897     223867.208659
44          sqft_living  185673.823688     223864.976253
45              grade#1  185673.823768     223862.809278
46          condition#1  185673.885285     223890.049936
47          condition#3  185673.902031     223902.951194

We can see that the validation error is minimum when we reach 16 features in the model (*condition#5	*). We stop the selection here even if the training error keeps getting smaller (overfitting).

Let's now calculate an estimation of the generalization error using *test_data*

In [17]:
greedy_algo_features_list = regression_greedy_algorithm['feature'].tolist()[:15] # select the first 30 features
test_error, _, _ = multiple_regression_model(train_data_2, test_data, greedy_algo_features_list)
print 'test error (RMSE) is: %s' %test_error

test error (RMSE) is: 191328.013498


The test error is getting smaller.

* Note 1: We could have used k-fold cross validation instead of a validation set.
* Note 2: Other greedy algorithms exist (e.g., backward stepwise, combining forward and backward steps)

Now, instead of searching over a discrete set of solutions using greedy algorithms, we can use another technique called regularization. We start with all possible features in the model and shrink the coefficients (weights). Two main regularization techniques exist, Ridge regression (a.k.a L2 regularization) and Lasso regression (a.k.a L1 regularization).

### 2.3 Ridge Regression

Ridge regression aims to avoid overfitting by adding a cost to the Residual Sum of Squares (RSS) term of standard least squares that depends on the 2-norm of the coefficients.  The result is penalizing fits with large coefficients.  The strength of this penalty, and thus the fit vs. model complexity, is controlled by a parameter alpha (here called "L2_penalty").

Let's test two models using alpha equal 1 and 10.

In [18]:
input_feature = train_data.columns.values.tolist() # list of column name
input_feature.remove('price')

for i in [1,10]:
    ridge = linear_model.Ridge(alpha = i, normalize = True) # initialize the model
    ridge.fit(train_data.as_matrix(columns = input_feature), train_data.as_matrix(columns = ['price'])) # fit the train data
    print 'test error (RMSE) is: %s' %mean_squared_error(test_data.as_matrix(columns = ['price']), 
                              ridge.predict(test_data.as_matrix(columns = [input_feature])))**0.5 # predict price and test error

test error (RMSE) is: 207440.450781
test error (RMSE) is: 292909.565734


Now the question is, how do we pick alpha to minimize the error?

Alpha is a measure of model complexity. So, as before, we could use a validation set to select alpha. However, that approach has a major disadvantage: it leaves fewer observations available for training. A better approach to overcome this issue is to use a cross-validation technique. It uses all of the training set in a smart way. k-fold cross-validation for example involves dividing the training set into k segments of roughly equal size. Similar to the validation set method, we measure the validation error with one of the segments designated as the validation set. The major difference is that we repeat the process k times. We then compute the average of the k validation errors, and use it as an estimate of the generalization error. We then select the alpha value that generate the smallest validation error. The best approximation occurs for validation sets of size 1, where k is equal to the number of observations. It is called leave-one-out cross validation. It is however computationally intensive.

In this example we'll use a ridge regression with an implemented cross-validation method from the scikit learn library. By default, it performs Generalized Cross-Validation, which is a form of efficient Leave-One-Out cross-validation.

In [19]:
ridgeCV = linear_model.RidgeCV(alphas = np.linspace(1.0e-10,1,num = 100), normalize = True, store_cv_values = True) # initialize the model
ridgeCV.fit(train_data.as_matrix(columns = input_feature), train_data.as_matrix(columns = ['price'])) # fit the train data
print 'best alpha is: %s' %ridgeCV.alpha_ # get the best alpha
print 'test error (RMSE) is: %s' %mean_squared_error(test_data.as_matrix(columns = ['price']), 
                              ridgeCV.predict(test_data.as_matrix(columns = [input_feature])))**0.5 # predict price and test error

best alpha is: 0.090909091
test error (RMSE) is: 191294.485755


Using every features in the dataset and a ridge regression model with an efficient  LOO cross-validation method yield a test error of 191294.

### 2.4 Lasso Regression

Lasso regression jointly shrinks coefficients to avoid overfitting, and implicitly performs feature selection by setting some coefficients exactly to 0 for sufficiently large penalty strength alpha (here called "L1_penalty"). In particular, lasso takes the RSS term of standard least squares and adds a 1-norm cost of the coefficients.

Let 's train multiple models using different alpha values and asses the test error.

In [20]:
for i in [0.01,0.1,1,250,500,1000]:
    lasso = linear_model.Lasso(alpha = i, normalize = True) # initialize the model
    lasso.fit(train_data.as_matrix(columns = input_feature), train_data.as_matrix(columns = ['price'])) # fit the train data
    print lasso.sparse_coef_.getnnz # number of non zero weights
    print 'test error (RMSE) is: %s' %mean_squared_error(test_data.as_matrix(columns = ['price']), 
                              lasso.predict(test_data.as_matrix(columns = [input_feature])))**0.5 # predict price and test error

<bound method csr_matrix.getnnz of <1x48 sparse matrix of type '<type 'numpy.float64'>'
	with 47 stored elements in Compressed Sparse Row format>>
test error (RMSE) is: 193326.649363
<bound method csr_matrix.getnnz of <1x48 sparse matrix of type '<type 'numpy.float64'>'
	with 44 stored elements in Compressed Sparse Row format>>
test error (RMSE) is: 193313.688132
<bound method csr_matrix.getnnz of <1x48 sparse matrix of type '<type 'numpy.float64'>'
	with 39 stored elements in Compressed Sparse Row format>>
test error (RMSE) is: 193229.877522
<bound method csr_matrix.getnnz of <1x48 sparse matrix of type '<type 'numpy.float64'>'
	with 10 stored elements in Compressed Sparse Row format>>
test error (RMSE) is: 217809.228424
<bound method csr_matrix.getnnz of <1x48 sparse matrix of type '<type 'numpy.float64'>'
	with 7 stored elements in Compressed Sparse Row format>>
test error (RMSE) is: 243769.338181
<bound method csr_matrix.getnnz of <1x48 sparse matrix of type '<type 'numpy.float64'>

As alpha increases, the number of features included in the model decrease. As in Ridge regression, we also can use cross-validation methods that select the best alpha to provide the best predictive accuracy. However, this technique tends to favor less sparse solutions and smaller alpha than optimal choice for feature selection.

In [21]:
lassoCV = linear_model.LassoCV(normalize = True) # initialize the model (alphas are set automatically)
lassoCV.fit(train_data.as_matrix(columns = input_feature), np.ravel(train_data.as_matrix(columns = ['price']))) # fit the train data
print 'best alpha is: %s' %lassoCV.alpha_ # get the best alpha
print 'number of non zero weigths is: %s' %np.count_nonzero(lassoCV.coef_) # number of non zero weights
print 'test error (RMSE) is: %s' %mean_squared_error(test_data.as_matrix(columns = ['price']), 
                              lassoCV.predict(test_data.as_matrix(columns = [input_feature])))**0.5 # predict price and test error

best alpha is: 36.9978057091
number of non zero weigths is: 28
test error (RMSE) is: 191627.546656


28 features remain in the model, yielding a test error of 191627.

### 2.5 k-Nearest Neighbors (NN) Regression

To finish, let's talk about another regression model, k-NN regression. The k-NN algorithm is used for estimating continuous variables. One such algorithm uses a weighted average of the k nearest neighbors, weighted by the inverse of their distance. It is the nonparamtric equivalent of ordinary least square regression.

In [22]:
# normalize the data
train_X = train_data.as_matrix(columns = input_feature)
scaler = preprocessing.StandardScaler().fit(train_X)
train_X_scaled = scaler.transform(train_X)
test_X = test_data.as_matrix(columns = [input_feature])
test_X_scaled = scaler.transform(test_X)

knn = neighbors.KNeighborsRegressor(n_neighbors=10, weights='distance') # initialize the model
knn.fit(train_X_scaled, train_data.as_matrix(columns = ['price'])) # fit the train data
print 'test error (RMSE) is: %s' %mean_squared_error(test_data.as_matrix(columns = ['price']), 
                              knn.predict(test_X_scaled))**0.5 # predict price and test error

test error (RMSE) is: 195427.612114


Similarly, the number of neighbor k is related to the complexity of the regression model. We can optimize the model accuracy by running cross-validation on the number of k-neighbors to include. 

## Conclusion

In this notebook I tried to give on overview of regression methods using a dataset of house sales. Obviously, many other methods exit for regression (e.g., Elastic Net, kernel regression, Bayesian regression). Model selection should be based on your data and application. 